Migrate Riddle of Fire to BuffWindow

Malpp opened the pull request:
Update the RoF module to BuffWindow as suggested in #1303 

![image](https://user-images.githubusercontent.com/17149559/152891708-b154a0ad-fc05-44c4-b0b9-26f04c9b01b2.png)
![image](https://user-images.githubusercontent.com/17149559/152891681-a246c454-ee90-4296-923c-520e440afe91.png)

Malpp committed:
Migrate Riddle of Fire to BuffWindow

Malpp committed:
Add profile picture

Malpp committed:
Changed missed brotherhood text

hintxiv commented on src/parser/jobs/mnk/changelog.tsx:
```diff
@@ -2,6 +2,11 @@ import CONTRIBUTORS from 'data/CONTRIBUTORS'
 export const changelog = [
+	{
+		date: new Date('2022-01-02'),
+		Changes: () => <>Migrated Riddle of Fire analyser to BuffWindow.</>,
```
Monk changelog uses imperative tense, could you rewrite this to say "migrate" instead of "migrated"? Also maybe just call it a "module" instead of "analyzer" since that's the leyperson's terminology for an xivanalysis module.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,199 @@
 import {Plural, Trans} from '@lingui/react'
-import {DataLink} from 'components/ui/DbLink'
-import {RotationTable} from 'components/ui/RotationTable'
+import {ActionLink, DataLink} from 'components/ui/DbLink'
```
No need to have both `ActionLink` and `DataLink` since the latter supercedes the former.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,199 @@
+class MasterfulBlitzEvaluator implements WindowEvaluator {
+	private _blitzSkills: Action[]
+	private _celestialRevolution: Action
+	private _masterfulBlitz: Action
```
The TypeScript `private` keyword replaces the need for prefixing variables with an underscore.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,199 @@
-		return this.casts.filter(event => event.action === this.data.actions.SIX_SIDED_STAR.id).length
+	private numberOfBlitzExpectedInWindow(windowIndex: number): number {
+		// 2 on odds and 1 on evens
+		return 2 - windowIndex % 2
```
This logic isn't robust enough. If there's a long downtime, or the user just makes a mistake with one of their PBs, you might flip your cycle (i.e. 1 blitz in window 1, 1 blitz in window 2, then 2, 1, 2, 1, 2, 1... from there). This means that only one window missed the requirement, but now we're going to flag every odd window because it doesn't follow the pattern.

I think the easiest solution for this is to just check the # of blitzes in the previous window - if the previous window had 2 blitzes, require 1; if it had 1, require 2.

ayyaruq replied:
I'm not sure that actually solves it, since downtime exists. It's possible user could have enough charges of PB that they can do 2 blitzes in 2 consecutive windows I think. Might be worth a time check since last PB or something.

Malpp replied:
Could I have something like: if the time between window 1 and window 2 is greater than 80 seconds (2 PB CDs), than I expect 2 blitzed in that window?

ayyaruq replied:
Sorry for the slow reply, yes I think that's a good approach.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,199 @@
 
-		this.riddle = undefined
-	}
+	blitzSkills = [this.data.actions.ELIXIR_FIELD, this.data.actions.RISING_PHOENIX, this.data.actions.PHANTOM_RUSH]
```
These are in `constants.ts` already, you probably want to just reuse that definition.

Malpp replied:
In the constants it's `ActionKey` but Im trying to use it with comparing `Action`.
This is the only spot I use it:
```ts
private countBlitzSkills(window: HistoryEntry<EvaluatedAction[]>): number {
    return window.data.filter(value => this.blitzSkills.includes(value.action)).length
}
```
Is there a way to do it?

hintxiv replied:
```ts
const blitzActions = BLITZ_SKILLS.map(key => this.data.actions[key].id)
```

ayyaruq replied:
MNK also has a `utilities.ts` with a `fillActions` function, so you can do like
```ts
const blitzActions = fillActions(BLITZ_SKILLS, this.data)
```

Malpp committed:
Better changelog message
Proper private variable names
Remove useless imports
Make number of excepted blitz more robust

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,208 @@
 import React from 'react'
 import {BLITZ_SKILLS} from './constants'
 import {DISPLAY_ORDER} from './DISPLAY_ORDER'
-import {fillActions} from './utilities'
```
You probably want this so you can fill your `blitzActions`.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,208 @@
+	2: SEVERITY.MAJOR,
+}
+
+const MEDITATION_DURING_ROF_SEVERITY = {
```
I'd maybe make this "bad gcds" or something, for example using anatman or form shift as well. I'd clean up the names of these vars a bit in general tho.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,208 @@
+
 const EXPECTED = {
 	GCDS: 11,
-	BLITZES: 1,
```
The only EXPECTED here is now GCDs, could prob just make this `const EXPECTED_GCDS = 11` and be done with it, although I'm not sure why this is needed since the expected GCD evaluator has the GlobalCooldown object yet also wants an expected field.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,208 @@
-		const playerFilter = filter<Event>().source(this.parser.actor.id)
-		this.addEventHook(playerFilter.type('statusApply').status(this.data.statuses.RIDDLE_OF_FIRE.id), this.onGain)
-		this.addEventHook(playerFilter.type('statusRemove').status(this.data.statuses.RIDDLE_OF_FIRE.id), this.onDrop)
+class BrotherhoodEvaluator implements WindowEvaluator {
```
At a high level, I'm not sure the value of this. It's not really a matter of if they used BH or not, but if it was wrapped into the window properly without drift. I was thinking more along the lines of a generic DriftEvaluator, so we could track the difference in timing and make sure the full duration of BH was inside RoF. Might be worth talking to #fb-monk people for feedback on this.

Malpp replied:
`DriftEvaluator` doesn't seem to exist? Unless it's something that would need to be created?

Right now since I'm tracking the action of doing BH inside the RoF window I guarantee the full during inside RoF.

I've noticed on my logs that I've been analyzing, sometimes I press BH before RoF because of timing difference, which shows up that I missed BH. I'm guessing the `DriftEvaluator` is what's going to handle that?

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,208 @@
+		this.addEvaluator(new LimitedActionsEvaluator({
+			expectedActions: [
+				{
+					action: this.data.actions.MEDITATION,
```
I'd add other non-damage GCDs to this, like Form Shift and Anatman.

ayyaruq reviewed (changes requested):
First pass, some thoughts on logic and what we're actually trying to track. Once this is sorted could do a bit of code cleanup (mostly layout/formatting). Very nice work tho!

Malpp committed:
Use fillActions
Better const names
Add Anatman and Form Shift to bad gcds during RoF

Malpp committed:
Merge remote-tracking branch 'upstream/endwalker' into endwalker2

# Conflicts:
#	src/data/CONTRIBUTORS.ts
#	src/parser/jobs/mnk/changelog.tsx

AzariahAeroborn commented:
@Malpp The lint error from the build is because you need to run `yarn extract` and commit the changes that will make to the locale files (due to changing the body of the suggestion content in commit 6d8df95).  

Also, if you can merge the endwalker branch back into your work branch (or rebase, whichever approach you prefer) and sort out the merge conflicts with the contributor & changelog files, we can poke the necessary reviewers to get the rest of it moving.

Malpp committed:
More robust Blitz + BH tracker based on window timings

Malpp committed:
Merge remote-tracking branch 'upstream/endwalker' into endwalker2

Malpp committed:
Update changelog

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,257 @@
+		}))
+	}
+
+	private numberOfSixSidedStarInWindow(window: HistoryEntry<EvaluatedAction[]>): number {
+		return window.data.filter(value => value.action.id === this.data.actions.SIX_SIDED_STAR.id).length
 	}
```
This is only called once and it's a one-liner, it doesn't need to be in its own function. You do this in a few places (e.g. `checkIfBrotherhoodIsInWindow`) - someone trying to contribute to this file in the future is going to have to jump around a lot to follow the logic.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,257 @@
+	constructor(blitzSkills: Array<Action['id']>, celestialRevolution: Action, masterfulBlitz: Action, brotherhood: Action) {
+		this.blitzSkills = blitzSkills
+		this.celestialRevolution = celestialRevolution
+		this.masterfulBlitz = masterfulBlitz
```
`masterfulBlitz` is assigned but never accessed, this seems like dead code.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,257 @@
-const EXPECTED = {
-	GCDS: 11,
-	BLITZES: 1,
+const TOTAL_GCD_DURING_ROF_SEVERITY = {
+	10: SEVERITY.MINOR,
+	9: SEVERITY.MEDIUM,
+	6: SEVERITY.MAJOR,
 }
```
A bit nitpicky, but you can combine all these severity consts into just one const:

```ts
const SEVERITIES = {
    TOTAL_GCD_DURING_ROF: { ... }
    ...
}

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,257 @@
-		if (this.riddle != null) {
-			this.riddle.end = endTime
+	private numberOfBrotherhoodsExpectedInWindow(windowIndex: number, windows: Array<HistoryEntry<EvaluatedAction[]>>): number {
+		// The first window should always have a Brotherhood
+		if (windowIndex === 0) {
+			return 1
+		}
```
If someone weaves BH > RoF in the opener, this will flag.

hintxiv commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,257 @@
+		}
 
-			this.history.push(this.riddle)
+		return 0
```
This doesn't really capture the nature of the mistake here - if a person does BH > RoF they'll get a 0/0 even though a normal person would consider this to be a BH + RoF window. If a person completely misaligns BH and it drops before RoF even goes up they'll also get 0/0 when we should be calling it 0/1. 

I'm okay with axing this logic altogether and saving it for a separate module (or a separate PR) but I don't think it should be merged as-is.

Malpp committed:
Move severities to const
Removed unused code
Simplify functions

Malpp committed:
Check if Brotherhood was used in 2 GCDs of RoF

Malpp committed:
Use brotherhood action instead of status

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+	BAD_GCD_DURING_ROF_SEVERITY: {
+		1: SEVERITY.MEDIUM,
+		2: SEVERITY.MAJOR,
+	},
```
Hypernit - these names are redundant. You're effectively writing `SEVERITY.FOO_SEVERITY`, they're all relative to RoF too, so I think `MISSED_GCDS`, `BAD_GCDS`, and `DRIFTED_BH` or something is clear enough.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
-	2: SEVERITY.MEDIUM,
-	5: SEVERITY.MAJOR,
-}
+const DOUBLE_PERFECT_BALANCE_IN_SECONDS = 80
```
This isn't needed.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
-		this.data = data
-		this.start = start
-		this.casts = []
+	constructor(blitzSkills: Array<Action['id']>, celestialRevolution: Action, brotherhood: Action) {
```
To make calling this easier, might wanna take some sort of `opts` object with these params as members. Also allows for easier defaults.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+					action="RISING_PHOENIX"/> and <ActionLink action="ELIXIR_FIELD"/> do more potency along with
+					AoE.</Trans>,
+			})
+		}
```
This can be a tiered severity. Since it's 1 use is major, if you set the tiers with `1: SEVERITY.MAJOR` it'll only show if they used CR, and less indentation is easier to read.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+		return {
+			format: 'table',
+			header: {
+				header: <ActionLink showName={false} action={'MASTERFUL_BLITZ'}/>,
```
Lets be consistent, unless there's a reason not to.
```suggestion
				header: <ActionLink showName={false} action="MASTERFUL_BLITZ"/>,
```

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
-	private history: Riddle[] = []
-	private riddle?: Riddle
-	private riddleHook?: EventHook<Events['action']>
+	private numberOfBlitzExpectedInWindow(windowIndex: number, windows: Array<HistoryEntry<EvaluatedAction[]>>): number {
```
Hypernit - can we shorten some of these names?

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+		// First window should always contain 2 blitz
+		if (windowIndex === 0) {
+			return 2
+		}
```
Still not keen on this, I think it's an unsafe assumption. For example, user goes into a gatekeeper fight second half and with BH on cooldown. You care about time since previous BH, not number of windows.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
-		const playerFilter = filter<Event>().source(this.parser.actor.id)
-		this.addEventHook(playerFilter.type('statusApply').status(this.data.statuses.RIDDLE_OF_FIRE.id), this.onGain)
-		this.addEventHook(playerFilter.type('statusRemove').status(this.data.statuses.RIDDLE_OF_FIRE.id), this.onDrop)
+		if (this.timeBetweenLastWindowInSeconds(windowIndex, windows) > DOUBLE_PERFECT_BALANCE_IN_SECONDS) {
```
As noted above, we don't really operate on seconds, I think you can just use the times directly with the action cooldown. This also doesn't guarantee they have 2 charges technically. Might be worth a timer hook on the cooldown itself or something here. Not too sure, might need to think about it.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+		const blitzSkillsInLastWindow = this.countBlitzSkills(windows[windowIndex - 1])
+		if (blitzSkillsInLastWindow === 2) {
+			return 1
+		}
```
You can fold this into a single line, or at least lose the const, same above tbh. Hypernit - slap a newline before the final return tho.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+			return 0
+		}
+		return (windows[windowIndex].start - windows[windowIndex].start) / 1000
+	}
```
I don't think this is needed.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
-		if (this.riddle != null) {
-			this.riddle.casts.push(event)
-		}
+	constructor(brotherhood: ActionRoot['BROTHERHOOD'], brotherhoodStartHistory: number[]) {
```
Same deal here, an options block might be better.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+		const possibleNumberOfBrotherhoods = Math.ceil(windows.length / 2)
+		const brotherhoodUsagesInsideRof = windows.map(previousValue => previousValue.data.filter(value => value.action === this.brotherhood))
+			.filter(value => value.length > 0).length
+		const missedBhInsideRof = possibleNumberOfBrotherhoods - brotherhoodUsagesInsideRof
```
Can we rename these? Something like `maxPossibleBH`, `nestedBH`, and `missedBH` are prob good enough. The BH in RoF var should probably be broken into multiple lines more sanely so it's easier to read.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+		return {
+			format: 'table',
+			header: {
+				header: <ActionLink showName={false} action={'BROTHERHOOD'}/>,
```
Same deal here.
```suggestion
				header: <ActionLink showName={false} action="BROTHERHOOD"/>,
```

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+			icon: this.brotherhood.icon,
+			tiers: SEVERITIES.BROTHERHOOD_OUTSIDE_ROF_SEVERITY,
+			value: missedBhInsideRof,
+			content: <Trans id="mnk.rof.suggestions.brotherhood.content">Missed {missedBhInsideRof} <ActionLink
+				action="BROTHERHOOD"/> inside <ActionLink action="RIDDLE_OF_FIRE"/> window.
+			</Trans>,
```
This should use proper i18n Plural to format correctly, and try to avoid line-breaking HTML tags and their params.

ayyaruq commented on src/parser/jobs/mnk/modules/RiddleOfFire.tsx:
```diff
@@ -1,198 +1,262 @@
+			content: <Trans id="mnk.rof.suggestions.brotherhood.content">Missed {missedBhInsideRof} <ActionLink
+				action="BROTHERHOOD"/> inside <ActionLink action="RIDDLE_OF_FIRE"/> window.
+			</Trans>,
+			why: <Trans id="mnk.rof.suggestions.brotherhood.why"><ActionLink
+				action="BROTHERHOOD"/> has a shorter duration than <ActionLink action="RIDDLE_OF_FIRE"/>.</Trans>,
```
This message doesn't really explain why they're being dinged. Also, this is a bit iffy to read, avoid line-breaking mid tag, you can start the line with ActionLink outright, and pop the closing Trans tag on the next line aligned with the `why` field.

ayyaruq reviewed (changes requested):
Stopping my review here, insane variable names aside, we need to rethink this window indexing stuff. At the current state I think it's easier to not use BuffWindow, which means we've definitely overcomplicated things because it shouldn't be easier to roll custom logic than using a core module designed for this purpose. Maybe we need to have a concrete list of goals for this PR, it seems to be handling some things but not others, and the way BH is done atm is unreliable. Happy to discuss although I'm taking a break from discord for a bit.

Malpp committed:
Remove brotherhood from RoF
Rename variables
Make Blitz more robust

Malpp committed:
Merge remote-tracking branch 'upstream/endwalker' into endwalker2

Malpp committed:
Merge remote-tracking branch 'upstream/endwalker' into endwalker2

Malpp committed:
Remove blitz counting
Ignore support actions
Remove CR suggestion

Malpp committed:
Run yarn extract

ayyaruq committed:
Cleanup some nits

ayyaruq committed:
Don't miss a spot

ayyaruq committed:
Don't miss the other spot either

ayyaruq reviewed (changes requested):
Just need another `yarn extract` and this is good to go. Thanks for your patience!

Malpp committed:
Run yarn extract
